test/ecto_diff_test.exs

```diff
@@ -614,6 +614,312 @@ defmodule EctoDiffTest do
+                 ]
+               }
+             } = diff
+    end
+
+    test "insert with empty has_many through" do
+      {:ok, pet} =
+        %{
+          name: "Spot",
+          resources: [
+            %{toys: []}
+          ]
+        }
+        |> Pet.new()
+        |> Repo.insert()
+
+      id = pet.id
+
+      {:ok, diff} = EctoDiff.diff(nil, pet)
+
+      assert %EctoDiff{
+               effect: :added,
+               primary_key: %{id: ^id},
+               changes: %{
+                 id: {nil, ^id},
+                 name: {nil, "Spot"},
+                 resources: [
+                   %EctoDiff{
+                     effect: :added,
+                     changes: %{
+                       pet_id: {nil, ^id}
+                     }
+                   }
+                 ]
+               }
+             } = diff
+    end
+
+    test "update with adding new has_many through records" do
+      {:ok, pet} = %{name: "Spot", resources: []} |> Pet.new() |> Repo.insert()
+
+      {:ok, updated_pet} =
+        pet
+        |> Pet.update(%{
+          resources: [
+            %{toys: [%{name: "ball", type: "play", quantity: 1}, %{name: "mouse", type: "play", quantity: 2}]}
+          ]
+        })
+        |> Repo.update()
+
+      updated_pet = Repo.preload(updated_pet, [:resources, :toys])
+
+      id = pet.id
+      [%{id: resource_id}] = updated_pet.resources
+      [ball_id, mouse_id] = Enum.map(updated_pet.toys, & &1.id)
+
+      {:ok, diff} = EctoDiff.diff(pet, updated_pet)
+
+      assert %EctoDiff{
+               effect: :changed,
+               primary_key: %{id: ^id},
+               changes: %{
+                 resources: [
+                   %EctoDiff{
+                     effect: :added,
+                     changes: %{
+                       pet_id: {nil, ^id},
+                       toys: [
+                         %EctoDiff{
+                           effect: :added,
+                           changes: %{}
+                         },
+                         %EctoDiff{
+                           effect: :added,
+                           changes: %{}
+                         }
+                       ]
+                     }
+                   }
+                 ],
+                 toys: [
+                   %EctoDiff{
+                     effect: :added,
+                     primary_key: %{id: ^ball_id},
+                     changes: %{
+                       id: {nil, ^ball_id},
+                       resource_id: {nil, ^resource_id},
+                       name: {nil, "ball"}
+                     }
+                   },
+                   %EctoDiff{
+                     effect: :added,
+                     primary_key: %{id: ^mouse_id},
+                     changes: %{
+                       id: {nil, ^mouse_id},
+                       resource_id: {nil, ^resource_id},
+                       name: {nil, "mouse"}
+                     }
+                   }
```

Comment:
:information_source: We should be able to assert that the EctoDiffs in these two arrays are the same:
```suggestion
                       toys: [ball_toy_diff, mouse_toy_diff]
                     }
                   }
                 ],
                 toys: [
                   %EctoDiff{
                     effect: :added,
                     primary_key: %{id: ^ball_id},
                     changes: %{
                       id: {nil, ^ball_id},
                       resource_id: {nil, ^resource_id},
                       name: {nil, "ball"}
                     }
                   } = ball_toy_diff,
                   %EctoDiff{
                     effect: :added,
                     primary_key: %{id: ^mouse_id},
                     changes: %{
                       id: {nil, ^mouse_id},
                       resource_id: {nil, ^resource_id},
                       name: {nil, "mouse"}
                     }
                   } = mouse_toy_diff
```

Reply:
Asserting this would only serve to prove that toys inside resources are the same as toys inside pets. This doesn't do much since toys in resources and toys in pets are diffed exactly the same, and toys in pets will be a copy of toys in resources (in nearly all cases). However, the test is easy enough, so I've added it.